File: backend/scripts/test-vector-mcp-store.php
<?php

declare(strict_types=1);

require dirname(__DIR__) . '/vendor/autoload.php';

use AgentTeam\Services\VectorMcpStore;

/**
 * VectorMcpStore unit tests — the store node's writer/reader. The MCP dispatch
 * is injected so no live vector server is needed. Verifies the unified `store`
 * (one batch call per file, items:[{text,metadata}]) and `find` contract, and
 * the mcp_qrant envelope decode (payload in content[0].text; errors ride INSIDE
 * the payload as {error:true,...}).
 */
$failures = 0;
$checks = 0;
function check(string $label, bool $cond): void
{
    global $failures, $checks;
    $checks++;
    echo ($cond ? "  PASS  " : "  FAIL  ") . $label . "\n";
    if (!$cond) {
        $failures++;
    }
}

/** Wrap a payload in the mcp_qrant envelope the dispatch callable returns. */
function envelope(array $payload): array
{
    return ['content' => [['type' => 'text', 'text' => json_encode($payload)]]];
}

// --- store: ONE `store` call carrying items[{text,metadata}] + provider/connection ---
$calls = [];
$dispatch = function (int $serverId, string $tool, array $args) use (&$calls) {
    $calls[] = compact('serverId', 'tool', 'args');
    return envelope(['stored' => 3, 'errors' => 0]);
};
$store = new VectorMcpStore($dispatch);
$res = $store->store(7, ['alpha', 'beta', '   ', 'gamma'], [
    'provider'   => 'qdrant',
    'connection' => ['mode' => 'local', 'path' => '/tmp/q'],
    'collection' => 'docs',
    'metadata'   => ['source' => 'a/x.pdf'],
]);

check('one store call total (batch per file)', count($calls) === 1);
check('uses the `store` tool', $calls[0]['tool'] === 'store');
check('targets the server id', $calls[0]['serverId'] === 7);
check('passes provider', $calls[0]['args']['provider'] === 'qdrant');
check('passes connection', ($calls[0]['args']['connection']['path'] ?? null) === '/tmp/q');
check('passes collection', $calls[0]['args']['collection'] === 'docs');
check('items skip the blank chunk (3 items)', count($calls[0]['args']['items']) === 3);
check('item carries text', $calls[0]['args']['items'][0]['text'] === 'alpha');
check('item carries metadata (provenance)', ($calls[0]['args']['items'][0]['metadata']['source'] ?? null) === 'a/x.pdf');
check('omits embedding when unset', !array_key_exists('embedding', $calls[0]['args']));
check('returns stored from payload', $res['stored'] === 3);
check('returns errors from payload', $res['errors'] === 0);
check('returns collection', $res['collection'] === 'docs');

// --- store: empty / all-blank chunks => no call ---
$calls2 = [];
$store2 = new VectorMcpStore(function (int $s, string $t, array $a) use (&$calls2) {
    $calls2[] = $a;
    return envelope(['stored' => 0, 'errors' => 0]);
});
$res2 = $store2->store(1, ['   ', ''], ['provider' => 'qdrant', 'connection' => [], 'collection' => 'c']);
check('all-blank chunks -> 0 stored, no call', $res2['stored'] === 0 && count($calls2) === 0);

// --- store: embedding forwarded when set ---
$ecalls = [];
(new VectorMcpStore(function (int $s, string $t, array $a) use (&$ecalls) {
    $ecalls[] = $a;
    return envelope(['stored' => 1, 'errors' => 0]);
}))->store(1, ['x'], ['provider' => 'pgvector', 'connection' => [], 'collection' => 'c', 'embedding' => 'hf:all-MiniLM-L6-v2']);
check('forwards embedding when set', ($ecalls[0]['embedding'] ?? null) === 'hf:all-MiniLM-L6-v2');

// --- store: payload error throws (error rides INSIDE the payload) ---
$threw = false;
try {
    (new VectorMcpStore(fn($s, $t, $a) => envelope(['error' => true, 'code' => 'store_failed', 'message' => 'boom'])))
        ->store(1, ['x'], ['provider' => 'qdrant', 'connection' => [], 'collection' => 'c']);
} catch (\RuntimeException $e) {
    $threw = $e->getMessage() === 'boom';
}
check('payload error throws', $threw);

// --- store: transport error (normalized {error:true}) throws ---
$threwT = false;
try {
    (new VectorMcpStore(fn($s, $t, $a) => ['error' => true, 'message' => 'HTTP 500']))
        ->store(1, ['x'], ['provider' => 'qdrant', 'connection' => [], 'collection' => 'c']);
} catch (\RuntimeException $e) {
    $threwT = $e->getMessage() === 'HTTP 500';
}
check('transport error throws', $threwT);

// --- find: builds args and maps results {text,metadata,score} ---
$fcalls = [];
$finder = new VectorMcpStore(function (int $s, string $t, array $a) use (&$fcalls) {
    $fcalls[] = compact('t', 'a');
    return envelope(['results' => [
        ['text' => 'solar power', 'metadata' => ['source' => 'energy.txt'], 'score' => 0.91],
        ['text' => 'no score', 'metadata' => [], 'score' => null],
    ]]);
});
$fr = $finder->find(5, 'renewable energy', [
    'provider' => 'qdrant', 'connection' => ['mode' => 'local'], 'collection' => 'docs',
]);
check('find uses the `find` tool', $fcalls[0]['t'] === 'find');
check('find passes query', $fcalls[0]['a']['query'] === 'renewable energy');
check('find default limit = 5', $fcalls[0]['a']['limit'] === 5);
check('find maps text', $fr['results'][0]['text'] === 'solar power');
check('find maps metadata', ($fr['results'][0]['metadata']['source'] ?? null) === 'energy.txt');
check('find maps score', $fr['results'][0]['score'] === 0.91);
check('find tolerates null score', $fr['results'][1]['score'] === null);

// --- find: payload error throws ---
$threwF = false;
try {
    (new VectorMcpStore(fn($s, $t, $a) => envelope(['error' => true, 'message' => 'find boom'])))
        ->find(1, 'q', ['provider' => 'qdrant', 'connection' => [], 'collection' => 'c']);
} catch (\RuntimeException $e) {
    $threwF = $e->getMessage() === 'find boom';
}
check('find payload error throws', $threwF);

// --- store: a response with NO {stored} field is a hard error (not silent 0) ---
// This is the class of bug a 307 redirect / wrong endpoint produced: an empty
// or non-MCP body that used to map to {stored:0} with no error.
$threwEmpty = false;
try {
    (new VectorMcpStore(fn($s, $t, $a) => []))   // {} — e.g. a redirect/non-MCP body
        ->store(1, ['x'], ['provider' => 'qdrant', 'connection' => [], 'collection' => 'c']);
} catch (\RuntimeException $e) {
    $threwEmpty = str_contains($e->getMessage(), 'no {stored}');
}
check('store: empty/non-MCP response throws (no silent stored:0)', $threwEmpty);

$threwOk = false;
try {
    (new VectorMcpStore(fn($s, $t, $a) => envelope(['ok' => true])))   // valid-looking, no `stored`
        ->store(1, ['x'], ['provider' => 'qdrant', 'connection' => [], 'collection' => 'c']);
} catch (\RuntimeException $e) {
    $threwOk = str_contains($e->getMessage(), 'no {stored}');
}
check('store: response without {stored} throws', $threwOk);

// --- find: a response with NO {results} field is a hard error (not "0 matches") ---
$threwFindEmpty = false;
try {
    (new VectorMcpStore(fn($s, $t, $a) => []))
        ->find(1, 'q', ['provider' => 'qdrant', 'connection' => [], 'collection' => 'c']);
} catch (\RuntimeException $e) {
    $threwFindEmpty = str_contains($e->getMessage(), 'no {results}');
}
check('find: empty/non-MCP response throws (no silent 0 matches)', $threwFindEmpty);

echo "\n$checks checks, $failures failures\n";
exit($failures === 0 ? 0 : 1);

File: backend_python/tests/unit/test_vector_mcp_store.py
"""Ported one-to-one from backend/scripts/test-vector-mcp-store.php (same
inputs, same expected values; `php backend/scripts/test-vector-mcp-store.php`
passes 28/28 against the current source, 2026-09-07).

VectorMcpStore is the store node's writer/reader. The MCP dispatch is
INJECTED (a callable) so no live vector server is needed -- this class does
no HTTP of its own; the real HTTP dispatch closure is built in
IngestionController (Task 6). Verifies the unified `store` (one batch call
per file, items:[{text,metadata}]) and `find` contract, and the mcp_qrant
envelope decode (payload in content[0].text; errors ride INSIDE the payload
as {error:true,...}).
"""
from __future__ import annotations

import json

import pytest

from app.agent_team.services.vector_mcp_store import VectorMcpStore


def _envelope(payload: dict) -> dict:
    """Wrap a payload in the mcp_qrant envelope the dispatch callable returns."""
    return {'content': [{'type': 'text', 'text': json.dumps(payload)}]}


# --- store: ONE `store` call carrying items[{text,metadata}] + provider/connection ---

def test_store_makes_one_batch_call_with_expected_shape():
    calls = []

    def dispatch(serverId, tool, args):
        calls.append({'serverId': serverId, 'tool': tool, 'args': args})
        return _envelope({'stored': 3, 'errors': 0})

    store = VectorMcpStore(dispatch)
    res = store.store(7, ['alpha', 'beta', '   ', 'gamma'], {
        'provider': 'qdrant',
        'connection': {'mode': 'local', 'path': '/tmp/q'},
        'collection': 'docs',
        'metadata': {'source': 'a/x.pdf'},
    })

    assert len(calls) == 1                                    # one store call total (batch per file)
    assert calls[0]['tool'] == 'store'                         # uses the `store` tool
    assert calls[0]['serverId'] == 7                           # targets the server id
    assert calls[0]['args']['provider'] == 'qdrant'            # passes provider
    assert calls[0]['args']['connection'].get('path') == '/tmp/q'  # passes connection
    assert calls[0]['args']['collection'] == 'docs'            # passes collection
    assert len(calls[0]['args']['items']) == 3                 # items skip the blank chunk (3 items)
    assert calls[0]['args']['items'][0]['text'] == 'alpha'     # item carries text
    assert calls[0]['args']['items'][0]['metadata'].get('source') == 'a/x.pdf'  # item carries metadata (provenance)
    assert 'embedding' not in calls[0]['args']                 # omits embedding when unset
    assert res['stored'] == 3                                  # returns stored from payload
    assert res['errors'] == 0                                  # returns errors from payload
    assert res['collection'] == 'docs'                         # returns collection


def test_store_all_blank_chunks_makes_no_call():
    calls2 = []

    def dispatch(serverId, tool, args):
        calls2.append(args)
        return _envelope({'stored': 0, 'errors': 0})

    store2 = VectorMcpStore(dispatch)
    res2 = store2.store(1, ['   ', ''], {'provider': 'qdrant', 'connection': {}, 'collection': 'c'})
    assert res2['stored'] == 0 and len(calls2) == 0


def test_store_forwards_embedding_when_set():
    ecalls = []

    def dispatch(serverId, tool, args):
        ecalls.append(args)
        return _envelope({'stored': 1, 'errors': 0})

    VectorMcpStore(dispatch).store(1, ['x'], {
        'provider': 'pgvector', 'connection': {}, 'collection': 'c', 'embedding': 'hf:all-MiniLM-L6-v2',
    })
    assert ecalls[0].get('embedding') == 'hf:all-MiniLM-L6-v2'


def test_store_payload_error_raises():
    def dispatch(serverId, tool, args):
        return _envelope({'error': True, 'code': 'store_failed', 'message': 'boom'})

    with pytest.raises(RuntimeError) as exc:
        VectorMcpStore(dispatch).store(1, ['x'], {'provider': 'qdrant', 'connection': {}, 'collection': 'c'})
    assert str(exc.value) == 'boom'


def test_store_transport_error_raises():
    def dispatch(serverId, tool, args):
        return {'error': True, 'message': 'HTTP 500'}   # normalized transport error, no envelope

    with pytest.raises(RuntimeError) as exc:
        VectorMcpStore(dispatch).store(1, ['x'], {'provider': 'qdrant', 'connection': {}, 'collection': 'c'})
    assert str(exc.value) == 'HTTP 500'


# --- find: builds args and maps results {text,metadata,score} ---

def test_find_builds_args_and_maps_results():
    fcalls = []

    def dispatch(serverId, tool, args):
        fcalls.append({'t': tool, 'a': args})
        return _envelope({'results': [
            {'text': 'solar power', 'metadata': {'source': 'energy.txt'}, 'score': 0.91},
            {'text': 'no score', 'metadata': {}, 'score': None},
        ]})

    finder = VectorMcpStore(dispatch)
    fr = finder.find(5, 'renewable energy', {
        'provider': 'qdrant', 'connection': {'mode': 'local'}, 'collection': 'docs',
    })

    assert fcalls[0]['t'] == 'find'                     # find uses the `find` tool
    assert fcalls[0]['a']['query'] == 'renewable energy'  # find passes query
    assert fcalls[0]['a']['limit'] == 5                 # find default limit = 5
    assert fr['results'][0]['text'] == 'solar power'    # find maps text
    assert fr['results'][0]['metadata'].get('source') == 'energy.txt'  # find maps metadata
    assert fr['results'][0]['score'] == 0.91            # find maps score
    assert fr['results'][1]['score'] is None            # find tolerates null score


def test_find_payload_error_raises():
    def dispatch(serverId, tool, args):
        return _envelope({'error': True, 'message': 'find boom'})

    with pytest.raises(RuntimeError) as exc:
        VectorMcpStore(dispatch).find(1, 'q', {'provider': 'qdrant', 'connection': {}, 'collection': 'c'})
    assert str(exc.value) == 'find boom'


# --- store/find: a response with NO {stored}/{results} field is a hard error
# (not a silent 0) -- this is the class of bug a 307 redirect / wrong
# endpoint produced: an empty or non-MCP body used to map to {stored:0} with
# no error. ---

def test_store_empty_non_mcp_response_raises():
    def dispatch(serverId, tool, args):
        return {}   # e.g. a redirect/non-MCP body

    with pytest.raises(RuntimeError) as exc:
        VectorMcpStore(dispatch).store(1, ['x'], {'provider': 'qdrant', 'connection': {}, 'collection': 'c'})
    assert 'no {stored}' in str(exc.value)


def test_store_response_without_stored_field_raises():
    def dispatch(serverId, tool, args):
        return _envelope({'ok': True})   # valid-looking, no `stored`

    with pytest.raises(RuntimeError) as exc:
        VectorMcpStore(dispatch).store(1, ['x'], {'provider': 'qdrant', 'connection': {}, 'collection': 'c'})
    assert 'no {stored}' in str(exc.value)


def test_find_empty_non_mcp_response_raises():
    def dispatch(serverId, tool, args):
        return {}

    with pytest.raises(RuntimeError) as exc:
        VectorMcpStore(dispatch).find(1, 'q', {'provider': 'qdrant', 'connection': {}, 'collection': 'c'})
    assert 'no {results}' in str(exc.value)


# --- connection default: PHP `(array) ($cfg['connection'] ?? [])` -- an
# unset connection is an empty PHP array, encoded as `[]` not `{}`. ---

def test_store_default_connection_is_empty_list_when_unset():
    calls = []

    def dispatch(serverId, tool, args):
        calls.append(args)
        return _envelope({'stored': 1, 'errors': 0})

    VectorMcpStore(dispatch).store(1, ['x'], {'provider': 'qdrant', 'collection': 'c'})
    assert calls[0]['connection'] == []
